src/iso8601-date.ts

```diff
@@ -0,0 +1,256 @@
+        if (!RE_YEAR.test(year)) {
+          throw InvalidIso8601Date;
+        }
+        if (peek() === END) {
+          state = ParseDateState.End;
+        } else {
+          consumeSeparator('-');
+          state = ParseDateState.Month;
+        }
+        break;
+      case ParseDateState.Month:
+        month = consume() + consume();
+        if (!RE_MONTH.test(month)) {
+          throw InvalidIso8601Date;
+        }
+
+        // YYYYMM is not a valid ISO-8601
+        // it requires a separator: YYYY-MM
+        if (hasSeparators === false && (peek() === END || peek() === 'T')) {
+          throw InvalidIso8601Date;
+        } else if (peek() === END) {
+          state = ParseDateState.End;
+        } else if (peek() === 'T') {
+          state = ParseDateState.Hour;
+        } else {
+          consumeSeparator('-');
+          state = ParseDateState.CalendarDate;
+        }
+        break;
+      case ParseDateState.CalendarDate:
+        date = consume() + consume();
+        if (!RE_DATE.test(date)) {
+          throw InvalidIso8601Date;
+        }
+
+        if (peek() === END) {
+          state = ParseDateState.End;
+        } else {
+          state = ParseDateState.Hour;
+        }
+        break;
+      case ParseDateState.Hour:
+        if (consume() !== 'T') {
+          throw InvalidIso8601Date;
+        }
+
+        hours = consume() + consume();
+        if (!RE_HOUR.test(hours)) {
+          throw InvalidIso8601Date;
+        }
+
+        if (peek() === END) {
+          state = ParseDateState.End;
+        } else if (['Z', '-', '+'].includes(peek() as string)) {
+          state = ParseDateState.Timezone;
+        } else {
+          consumeSeparator(':');
+          state = ParseDateState.Minute;
+        }
+        break;
+      case ParseDateState.Minute:
+        minutes = consume() + consume();
+        if (!RE_MINUTE_SECOND.test(minutes)) {
+          throw InvalidIso8601Date;
+        }
+
+        if (peek() === END) {
+          state = ParseDateState.End;
+        } else if (['Z', '-', '+'].includes(peek() as string)) {
+          state = ParseDateState.Timezone;
+        } else {
+          consumeSeparator(':');
+          state = ParseDateState.Second;
+        }
+        break;
+      case ParseDateState.Second:
+        seconds = consume() + consume();
+        if (!RE_MINUTE_SECOND.test(seconds)) {
+          throw InvalidIso8601Date;
+        }
+
+        if (peek() === END) {
+          state = ParseDateState.End;
+        } else {
+          state = ParseDateState.Timezone;
+        }
+        break;
+      case ParseDateState.Timezone:
+        timezone = consume();
+
+        if (timezone === 'Z') {
+          if (peek() !== END) {
+            throw InvalidIso8601Date;
+          }
+          state = ParseDateState.End;
+        } else if (['-', '+'].includes(timezone)) {
+          state = ParseDateState.TimezoneHour;
+        } else {
+          throw InvalidIso8601Date;
+        }
```

Comment:
This logic should probably be updated in accordance with https://github.com/MetaMask/SIPs/pull/152